### PR TITLE
Store source file in TASTY attributes

### PR DESCRIPTION
**Pull request by nicolasstucki**
Add source file to TASTy attributes. This is a first step towards removing the `@SourceFile` annotation.

#### Release notes
Tools that read TASTy need to know that the source file must be unpickled from TASTy attributes.

### CODE REVIEW

File: tasty/src/dotty/tools/tasty/TastyFormat.scala [hunk 1]
```diff
@@ -265,13 +266,20 @@ All elements of a position section are serialized as Ints
...
 ```none
-  Comment       = UTF8 LongInt              // Raw comment's bytes encoded as UTF-8, followed by the comment's coordinates.
+  Comment       = Utf8 LongInt              // Raw comment's bytes encoded as UTF-8, followed by the comment's coordinates.
 ```
 
 Standard Section: "Attributes" Attribute*
 ```none
   Attribute     = SCALA2STANDARDLIBRARYattr
                   EXPLICITNULLSattr
+                  SOURCEFILEattr Utf8
```
Comment (bishabosha) [inline tasty/src/dotty/tools/tasty/TastyFormat.scala:276]: should this be raw utf8 or a nametable reference?

File: tasty/src/dotty/tools/tasty/TastyFormat.scala [hunk 2]
```diff
@@ -273,6 +274,13 @@ Standard Section: "Attributes" Attribute*
 ```none
   Attribute     = SCALA2STANDARDLIBRARYattr
                   EXPLICITNULLSattr
+                  SOURCEFILEattr Utf8Ref
+```
+
+Note: Attribute tags are grouped into 2 categories that determine what follows, and thus allow to compute the size of the tagged tree in a generic way.
+```none
+  Attribute Category 1 (tags 1-64)  :  tag
+  Attribute Category 2 (tags 65-128):  tag Utf8Ref
```
Comment (nicolasstucki) [inline tasty/src/dotty/tools/tasty/TastyFormat.scala:283]: I wonder if I should generalize this category to `tag Length Utf8Ref*`. In this case we could define the source attribute as `SOURCEFILEattr 1 Utf8Ref`.

Comment (bishabosha) [reply tasty/src/dotty/tools/tasty/TastyFormat.scala:283]: e.g. for macro arguments?

Comment (nicolasstucki) [reply tasty/src/dotty/tools/tasty/TastyFormat.scala:283]: Yes. But that is one we should not add yet as we do not have a sound spec for it.

File: tasty/src/dotty/tools/tasty/TastyFormat.scala (hunk 1 above)
Comment (nicolasstucki) [reply tasty/src/dotty/tools/tasty/TastyFormat.scala]: Yes

File: compiler/src/dotty/tools/dotc/core/tasty/AttributeUnpickler.scala [hunk 3]
```diff
@@ -3,19 +3,30 @@ package core.tasty
...
 
     while !isAtEnd do
-      booleanTags += readByte()
+      val tag = readByte()
+      if isBooleanAttrTag(tag) then
+        booleanTags += tag
+      else if isStringAttrTag(tag) then
+        val utf8Ref = readNameRef()
+        val value = nameAtRef(utf8Ref).toString
+        stringTagValue += tag -> value
```
[the remark this answers is not in the document]
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/core/tasty/AttributeUnpickler.scala:25]: Maybe I should add to the spec the fact that attribute tags cannot be repeated. I could also specify that they must be in order to make it reproducible.

**Review by bishabosha — APPROVED** [review]

**Pull request merged**